devicetypes/stelpro/stelpro-ki-zigbee-thermostat.src/stelpro-ki-zigbee-thermostat.groovy

```diff
@@ -0,0 +1,687 @@
+    return pollCalled()
+}
+
+def pollCalled() {
+    delayBetween([
+    			updateWeather(),
+                zigbee.readAttribute(0x201, 0x0000),	//Read Local Temperature
+                zigbee.readAttribute(0x201, 0x0008),	//Read PI Heating State
+    			zigbee.readAttribute(0x201, 0x0012),	//Read Heat Setpoint
+                zigbee.readAttribute(0x201, 0x001C),	//Read System Mode
+                my_readAttribute(0x201, 0x401C, ["mfgCode": "0x1185"]),	//Read Manufacturer Specific Setpoint Mode
+                zigbee.readAttribute(0x204, 0x0000),	//Read Temperature Display Mode
+                zigbee.readAttribute(0x204, 0x0001)		//Read Keypad Lockout
+        ], 200)
+    sendEvent( name: 'change', value: 0 )
+}
+
+
+
+def getTemperature(value) {
+	if (value != null) {
+    	log.debug("value $value")
+		def celsius = Integer.parseInt(value, 16) / 100
+		if (getTemperatureScale() == "C") {
+			return celsius
+		}
+        else {
+			return Math.round(celsiusToFahrenheit(celsius))
+		}
+	}
+}
+
+def refresh() {
+    poll()
+}
+
+def quickSetHeat(degrees) {
+    sendEvent( name: 'change', value: 1 )
+    setHeatingSetpoint(degrees, 0)
+}
+
+def setHeatingSetpoint(preciseDegrees, delay = 0) {
+	if (preciseDegrees != null) {
+		def temperatureScale = getTemperatureScale()
+		def degrees = new BigDecimal(preciseDegrees).setScale(1, BigDecimal.ROUND_HALF_UP)
+
+		log.debug "setHeatingSetpoint({$degrees} ${temperatureScale})"
+        
+        sendEvent(name: "heatingSetpoint", value: degrees, unit: temperatureScale, data: [heatingSetpointRange: heatingSetpointRange])
+    	sendEvent(name: "thermostatSetpoint", value: degrees, unit: temperatureScale, data: [thermostatSetpointRange: thermostatSetpointRange])
+        
+        def celsius = (getTemperatureScale() == "C") ? degrees : (fahrenheitToCelsius(degrees) as Float).round(2)
+        delayBetween([
+        	zigbee.writeAttribute(0x201, 0x12, 0x29, hex(celsius * 100)),
+        	zigbee.readAttribute(0x201, 0x12),	//Read Heat Setpoint
+            zigbee.readAttribute(0x201, 0x08),	//Read PI Heat demand
+    	], 100)
+	}
+}
+
+def setCoolingSetpoint(degrees) {
+	log.trace "${device.displayName} does not support cool setpoint"
+}
+
+def quickSetOutTemp(degrees) {
+    setOutdoorTemperature(degrees, 0)
+}
+
+def setOutdoorTemperature(degrees, delay = 0) {
+    setOutdoorTemperature(degrees.toDouble(), delay)
+}
+
+def setOutdoorTemperature(Double degrees, Integer delay = 0) {
+    def p = (state.precision == null) ? 1 : state.precision
+    Integer tempToSend
+    def tempToSendInString
+    
+    def celsius = (getTemperatureScale() == "C") ? degrees : (fahrenheitToCelsius(degrees) as Float).round(2)
+
+    if (celsius < 0) {
+        tempToSend = (celsius*100) + 65536
+    }
+    else {
+    	tempToSend = (celsius*100)
+    }
+    tempToSendInString = zigbee.convertToHexString(tempToSend, 4)
+    my_writeAttribute(0x201, 0x4001, 0x29, tempToSendInString, ["mfgCode": "0x1185"])
```

Comment:
You can use `zigbee.writeAttribute` to do Manufacturer specific writes similar to `zigbee.readAttribute`

Reply:
Is this new?
Manufacturer specific had to be written separately before.